keps/4136-admission-fair-sharing/README.md

```diff
@@ -0,0 +1,245 @@
+	UsageBasedFairSharing AdmissionMode = UsageBasedFairSharing 
+
+NoFairSharing AdmissionMode = NoFairSharing	
+)
+
+type AdmissionScope struct {
+    AdmissionMode AdmissionMode
+}
+```
+
+* When selecting candidates for admission groups all workloads from LQ to CQ or the
+topmost Cohort that is marked with AdmissionScope. Then sort them using criterias:
+
+- Usage vector  built from ConsumedResources from TopCohort to LQ
+- Priority
+- Timestamp
+
+The usage vector will be sum of the `ConsumedResources` weighted according to `resourceWeights`
+(mentioned later in the KEP)
+
+Let’s look at a couple scenarios:
+
+1. AdmissionScope at CQ, CQ queueing policy is FIFO, 3 LQ pointing to CQ. Kueue considers
+all CQ resources and potentially borrowed resources as “shared” resources and fair sharing
+is applied to all workloads.
+Kueue sorts the workloads by their LQ usage (if mode is `UsageBasedFairSharing`), priority and
+timestamp and tries to admit the first one from the list. Other workloads are not attempted
+until the first one is not admitted.
+
+2.  AdmissionScope at CQ, CQ queueing policy is BestEffort, 3 LQ pointing to CQ. 
+Kueue considers all CQ resources and potentially borrowed resources as “shared” resources and
+fair sharing is applied to all workloads.
+
+Kueue sorts the workloads by their LQ usage (if mode is `UsageBasedFairSharing`), priority and
+timestamp and tries to admit the first one from the list. If it fails and the second, third
+or following is possible then that workload is admitted, under condition that it might get preempted.
+
+3. AdmissionScope at Cohort level - Kueue operates in a mixed mode. Inside CQ workloads are
+selected according to their AdmissionMode (if specified). If a workload fits entirely into
+nominal quota, then it is admitted immediately, if not it goes into cohort-level fair sharing.
+For Cohort we select all the “sticking out” workloads, and sort them by their CQ usage, priority
+and timestamp. Kueue attempts to admit the first workload from the list of sticking-out
+(workload + current_usage > resources), just like if it was one big strict FIFO queue. 
+For multi-level hierarchy under one AdmissionScope we would treat underlying Cohorts as Fifo CQs.
+
+
+* Additionally there will be the resource usage calculation loop. The frequency of the
+calculation will be controlled globally in Kueue’s config file. Accounting would be done
+using something like geometric average:
+
+usage_sum = (1-A) * previous_usage_sum + A * current_usage.  
+
+The value will be stored in FairSharingStatus for all LQ, CQ, and Cohorts.
+
+The user will be able to configure the decaying factor A in Kueue’s config file by specifying
+the half decay time - after what time the current shared usage will decay to half of its original value.
+
+A = 1 - 0.5 ^ (sampling/half_decay)
+
+* Configuration will sit in FairSharing stuct in Kueue config. There will be the following modifications:
+
```

Comment:
Can we agree on more concrete shape of the config API, as the follow-up to this discussion: https://github.com/kubernetes-sigs/kueue/pull/4687#discussion_r2013886300

Reply:
What would be your proposal?

Reply:
I'd propose to embed `usageHalfLifeDecayTime`, `usageSamplingInterval`, `resourceWeights` into a struct called `AdmissionFairSharing` and make it a pointer. Then the presence of non-nil value of this pointer could determine if the Admission mode is switch on/off, so there's no possibility of having invalid state of the configuration. 

Ideally, in the next version of API I think we should drop the `Enable` field and migrate towards a struct `PreemptionFairSharing` which presence would determine if it's switched on/off

Reply:
Ok.

Reply:
Since you agree can we remove `modes` field?

Reply:
Done.